facade/src/main/java/org/jboss/pnc/facade/deliverables/DeliverableAnalyzerManagerImpl.java

```diff
@@ -782,39 +758,29 @@ private void prefetchTargetRepos(Collection<Build> builds) {
         }
 
         private void prefetchBrewArtifacts(Collection<Build> builds) {
-            Set<IdentifierShaRepo> queries = builds.stream()
+            Set<IdentifierSha256> identifierSha256Set = builds.stream()
                     .filter(b -> b.getBuildSystemType() == BuildSystemType.BREW)
                     .flatMap(this::prefetchBrewBuild)
                     .collect(Collectors.toSet());
 
-            Set<org.jboss.pnc.model.Artifact.IdentifierSha256> identifierSha256Set = queries.stream()
-                    .map(IdentifierShaRepo::getIdentifierSha256)
-                    .collect(Collectors.toSet());
             Set<org.jboss.pnc.model.Artifact> artifacts = artifactRepository
                     .withIdentifierAndSha256(identifierSha256Set);
-            for (org.jboss.pnc.model.Artifact artifact : artifacts) {
-                IdentifierShaRepo key = new IdentifierShaRepo(
-                        artifact.getIdentifierSha256(),
-                        artifact.getTargetRepository());
-                if (queries.contains(key)) {
-                    brewCache.put(key, artifact);
-                }
-            }
-            log.debug("Preloaded {} brew artifacts to cache.", targetRepositoryCache.size(), brewCache.size());
-        }
-
-        public Stream<IdentifierShaRepo> prefetchBrewBuild(Build build) {
-            String path = getKojiPath(build);
-            TargetRepository targetRepository = targetRepositoryCache.get(path);
-            if (targetRepository == null) {
-                return Stream.empty();
-            } else {
-                return build.getArtifacts()
-                        .stream()
-                        .peek(this::assertBrewArtifacts)
-                        .map(a -> mapBrewArtifact(a, build.getBrewNVR(), targetRepository, user))
-                        .map(a -> new IdentifierShaRepo(a.getIdentifierSha256(), targetRepository));
-            }
+            // Search for all artifacts with the provided SHA-256 and identifiers.
+            // If more than one artifact is found for the same SHA-256 and identifier (should not happen for Maven
+            // artifacts!), find a best match.
+            Map<IdentifierSha256, List<org.jboss.pnc.model.Artifact>> groupedByIdentifierSha256 = artifacts.stream()
+                    .collect(Collectors.groupingBy(org.jboss.pnc.model.Artifact::getIdentifierSha256));
+            groupedByIdentifierSha256.forEach(
+                    (key, matchedArtifacts) -> brewCache.put(key, getBestMatchingArtifact(matchedArtifacts).get()));
```

Comment:
for non-imported artifacts, the best match should not be IMPORTED, I think

Reply:
In this case, we are looking for identifier+sha256 of artifacts found in Brew. Which means that those are not built in PNC because BF gives precedence to PNC over Brew. Ideally those are all imported (but might not have all IMPORTED statuses)...am I missing some corner case?

Reply:
@janinko I have separated the behavior accordingly after chats.